### ISSUE TITLE
Technical Risk Assessment for Continuous integration

### ISSUE THREAD

**Issue by maxime-rainville** [silverstripe/github-actions-ci-cd#4]
Before we can widely deploy this tool, we need to measure to perform a technical risk assessment to minimise the chances that it might introduce vulnerability to our code base or our CI/CI pipeline.

# Acceptance
- [x] Risks have been identified
- [x] Impact and likelihood ratings have been recorded for each risk
- [x] Controls have been identified to manage risks 

# Scope
There's lots of scope for using GitHub actions with Silverstripe CMS. This initial phase will be limited to:
- Risks inherent with using GitHub actions
- Risks inherent with using GitHub actions for continuous integration

Technical Risk will be reassesses when using GitHub Actions is being contemplated for new use cases.

# Related resources
- [How to use GitHub Actions with security in mind](https://githubuniverse.com/content-library/how-to-use-github-actions-with-security-in-mind/)

**Comment by maxime-rainville** [comment]
We are working on a [Google Spread sheet](https://docs.google.com/spreadsheets/d/1iDnHU9Ojm3dC5PZWkY0s_OA1dmcwWiM0-4qi-XR6WBQ/edit#gid=0) (limited to Silverstripe Staff for now).

**Comment by maxime-rainville** [comment]
Just did a review of the proposed controls:
- For "We loose the ability to update/maintian builds", I would like a control like "write proper docs/API on how to use each workflow and custom actions" or "adopt best practices for managing GitHub actions".
- For "Vulnerability is introduced into a dependant action" can we a have some control like "we have process for reviewing and approving actions before including them in our workflow"
  - Locking actions to specific SHAs seems a bit counter intuitive since we wouldn't do that for any other PHP/JS dependencies. Then maybe the GitHub actions ecosystem is perhaps not as mature as the PHP/JS ecosystem. For example, I don't think GH actions as any concepts like lock files.
- For "Builds don't fail when they are meant to" seems like a "best practices" control would address that. There's also a [bash flag you can put in to force hard failures](https://intoli.com/blog/exit-on-errors-in-bash-scripts/) which might work in GH actions as well.
- Can we document all the proposed control in the Control tab?
![image.png](https://images.zenhubusercontent.com/5ac42fb74b5806bc2bcd8e13/a36910ab-0029-4f2a-b61e-9bf3ac627881)

**Comment by emteknetnz** [comment]
> For "Vulnerability is introduced into a dependant action" can we a have some control like "we have process for reviewing and approving actions before including them in our workflow"

No, this won't do anything to prevent this risk form occurring.  The risk comes from a 3rd party action, on an existing workflow that has been through peer review, having malicious code being added to it and updating a stable tag so that includes the malicious code.

The purpose of using a pinned sha is that we'll always fetch the exact same code, meaning that it can't fetch any future malicious commits.

**Comment by emteknetnz** [comment]
> Can we document all the proposed control in the Control tab?

Is that's just literally copy-pasting the content in the 'Additional Controls being implemented' column?

**Comment by emteknetnz** [comment]
I've updated as per the other recommendations you had

**Comment by maxime-rainville** [comment]
Risk register is up to date. We'll just need to create matching cards to go along with it.
